### ISSUE TITLE
Server Crashes Cause Graph Corruption

### ISSUE THREAD

**Issue by Kneelawk** [Kneelawk/GraphLib#10]
# The Issue
Currently, if the server crashes, it can shutdown without saving the latest graph and chunk updates. This means then that any wires that relied on these graph connections will no longer be connected correctly when the server is started back up.

# Potential Fix
A potential fix for this is to switch the `simple` graph implementation to be being saved every tick, only saving the parts that have been updated since the last save.

## Potential Drawbacks
If too many changes are made to graphs every tick, this could have significant TPS performance impacts. However, this kind of situation is unlikely.

However, what might be better is to have multiple `BlockGraph` implementations, allowing mods that update graphs frequently to choose to save their data only at regular intervals, while allowing mods that update graphs infrequently to save their data after every update, making corruption less likely (#11).

**Comment by MattiDragon** [comment]
Would it be possible to make sure that the graphs save together with the chunks? Then blocks and graphs never desync unless the crash happens during saving

**Comment by Kneelawk** [comment]
I've been attempting to do something like this, saving graph chunks and contained graphs when the world chunks unload. But it still doesn't seem to be working. I could probably try and find a precise injection point for most reliable saving, but I'm not sure how easy that would be.

Point-of-Interest saving is done the way I suggested it and it doesn't seem to run into the same desync issues that I'm currently encountering, which is why I suggested doing things this way.
